File: src/Importer.php
<?php

declare(strict_types=1);

final class Importer
{
    private const MONTHS = [
        ['E', 1, 'فروردین'],
        ['F', 2, 'اردیبهشت'],
        ['G', 3, 'خرداد'],
        ['H', 4, 'تیر'],
        ['I', 5, 'مرداد'],
        ['J', 6, 'شهریور'],
        ['K', 7, 'مهر'],
        ['L', 8, 'آبان'],
        ['M', 9, 'آذر'],
        ['N', 10, 'دی'],
        ['O', 11, 'بهمن'],
        ['P', 12, 'اسفند'],
    ];

    public function __construct(private readonly PDO $pdo, private readonly string $basePath)
    {
    }

    /**
     * @return array<string, int>
     */
    public function importAll(): array
    {
        Schema::migrate($this->pdo);

        $this->pdo->beginTransaction();
        try {
            Schema::reset($this->pdo);
            $this->importInnovationCenter($this->basePath . '/Innovation Center.xlsx');
            $this->importCharges($this->basePath . '/CHARGE.xlsx');
            $this->importFinance($this->basePath . '/finance.xlsx');
            $this->insert('import_runs', [
                'source_files' => json_encode(
                    ['Innovation Center.xlsx', 'CHARGE.xlsx', 'finance.xlsx'],
                    JSON_UNESCAPED_UNICODE
                ),
            ]);
            $this->pdo->commit();
        } catch (Throwable $exception) {
            $this->pdo->rollBack();
            throw $exception;
        }

        return [
            'members' => $this->count('members'),
            'teams' => $this->count('teams'),
            'lockers' => $this->count('lockers'),
            'plans' => $this->count('plans'),
            'charges' => $this->count('charges'),
            'transactions' => $this->count('transactions'),
            'warnings' => $this->count('import_warnings'),
        ];
    }

    private function importInnovationCenter(string $path): void
    {
        $reader = new XlsxReader($path);
        $file = basename($path);

        for ($row = 4; $row <= $reader->maxRow('Members'); $row++) {
            $fullName = $this->clean($reader->value('Members', "C{$row}"));
            if ($fullName === '') {
                continue;
            }
            $this->insert('members', [
                'row_number' => $row,
                'code' => $this->clean($reader->value('Members', "B{$row}")),
                'full_name' => $fullName,
                'team_name' => $this->clean($reader->value('Members', "D{$row}")),
                'desks' => $this->clean($reader->value('Members', "E{$row}")),
                'lockers' => $this->clean($reader->value('Members', "F{$row}")),
                'power_strips' => $this->clean($reader->value('Members', "G{$row}")),
                'phone' => $this->clean($reader->value('Members', "H{$row}")),
                'national_id' => $this->clean($reader->value('Members', "I{$row}")),
                'notes' => $this->clean($reader->value('Members', "J{$row}")),
                'source_file' => $file,
                'source_sheet' => 'Members',
            ]);
        }

        for ($row = 6; $row < 20; $row++) {
            $name = $this->clean($reader->value('Teams', "B{$row}"));
            $leader = $this->clean($reader->value('Teams', "C{$row}"));
            if ($name === '' && $leader === '') {
                continue;
            }
            $this->insert('teams', [
                'row_number' => $row,
                'name' => $name,
                'leader' => $leader,
                'phone' => $this->clean($reader->value('Teams', "D{$row}")),
                'desk_count' => $this->parseInt($reader->value('Teams', "E{$row}")),
                'lockers' => $this->clean($reader->value('Teams', "F{$row}")),
                'power_strips' => $this->clean($reader->value('Teams', "G{$row}")),
                'joined_at' => $this->parseJalaliDate($reader->value('Teams', "H{$row}")),
                'warning' => $this->clean($reader->value('Teams', "I{$row}")),
                'notes' => $this->clean($reader->value('Teams', "J{$row}")),
                'source_file' => $file,
                'source_sheet' => 'Teams',
            ]);
        }

        for ($row = 6; $row <= $reader->maxRow('lockers'); $row++) {
            $lockerNumber = $this->parseInt($reader->value('lockers', "A{$row}"));
            if ($lockerNumber === null) {
                continue;
            }
            $deliveredAt = $this->parseJalaliDate($reader->value('lockers', "D{$row}"));
            if ($deliveredAt !== '' && !$this->isValidJalaliDate($deliveredAt)) {
                $this->warn($file, 'lockers', $row, 'تاریخ تحویل کمد معتبر به نظر نمی‌رسد.', [
                    'delivered_at' => $deliveredAt,
                    'locker_number' => $lockerNumber,
                ]);
            }
            $this->insert('lockers', [
                'locker_number' => $lockerNumber,
                'status' => $this->clean($reader->value('lockers', "B{$row}")),
                'assigned_to' => $this->clean($reader->value('lockers', "C{$row}")),
                'delivered_at' => $deliveredAt,
                'key_number' => $this->clean($reader->value('lockers', "E{$row}")),
                'spare_key' => $this->clean($reader->value('lockers', "F{$row}")),
                'notes' => '',
                'source_file' => $file,
                'source_sheet' => 'lockers',
            ]);
        }

        for ($row = 6; $row <= $reader->maxRow('plans'); $row++) {
            $planNumber = $this->parseInt($reader->value('plans', "A{$row}"));
            $title = $this->clean($reader->value('plans', "C{$row}"));
            if ($planNumber === null && $title === '') {
                continue;
            }
            $this->insert('plans', [
                'plan_number' => $planNumber,
                'status' => $this->clean($reader->value('plans', "B{$row}")),
                'title' => $title,
                'proposed_budget' => $this->parseInt($reader->value('plans', "D{$row}")),
                'cost_type' => $this->clean($reader->value('plans', "E{$row}")),
                'schedule' => $this->clean($reader->value('plans', "F{$row}")),
                'notes' => '',
                'source_file' => $file,
                'source_sheet' => 'plans',
            ]);
        }
    }

    private function importCharges(string $path): void
    {
        $reader = new XlsxReader($path);
        $file = basename($path);

        foreach ($reader->sheetNames() as $sheet) {
            $chargeRate = $this->parseInt($reader->value($sheet, 'T1'));
            $rentRate = $this->parseInt($reader->value($sheet, 'T3'));
            for ($row = 6; $row <= $reader->maxRow($sheet); $row++) {
                $teamName = $this->clean($reader->value($sheet, "B{$row}"));
                $leader = $this->clean($reader->value($sheet, "C{$row}"));
                if ($teamName === '' && $leader === '') {
                    continue;
                }

                $note = $this->clean($reader->value($sheet, "Q{$row}"));
                foreach (self::MONTHS as [$column, $monthIndex, $monthName]) {
                    $amount = $this->parseInt($reader->value($sheet, "{$column}{$row}"));
                    if ($amount === null) {
                        continue;
                    }
                    $this->insert('charges', [
                        'fiscal_year' => $sheet,
                        'team_name' => $teamName,
                        'leader' => $leader,
                        'desk_count' => $this->parseInt($reader->value($sheet, "D{$row}")),
                        'month_index' => $monthIndex,
                        'month_name' => $monthName,
                        'amount' => $amount,
                        'note' => $note,
                        'charge_rate' => $chargeRate,
                        'rent_rate' => $rentRate,
                        'source_file' => $file,
                        'source_sheet' => $sheet,
                    ]);
                }
            }
        }
    }

    private function importFinance(string $path): void
    {
        $reader = new XlsxReader($path);
        $file = basename($path);

        foreach ($reader->sheetNames() as $sheet) {
            $this->insert('financial_batches', [
                'sheet_name' => $sheet,
                'petty_cash_holder' => $this->clean($reader->value($sheet, 'I1')),
                'petty_cash_number' => $this->clean($reader->value($sheet, 'K1')),
                'previous_balance' => $this->parseInt($reader->value($sheet, 'I2')),
                'new_deposit' => $this->parseInt($reader->value($sheet, 'I3')),
                'total_balance' => $this->parseInt($reader->value($sheet, 'I4')),
                'received_at' => $this->parseJalaliDate($reader->value($sheet, 'K2')),
                'from_date' => $this->parseJalaliDate($reader->value($sheet, 'K3')),
                'to_date' => $this->parseJalaliDate($reader->value($sheet, 'K4')),
                'source_file' => $file,
            ]);
            $batchId = (int) $this->pdo->lastInsertId();

            for ($row = 6; $row <= 27; $row++) {
                $rowNumber = $this->clean($reader->value($sheet, "A{$row}"));
                $description = $this->clean($reader->value($sheet, "F{$row}"));
                $amount = $this->parseInt($reader->value($sheet, "I{$row}"));
                $notes = $this->clean($reader->value($sheet, "J{$row}"));
                if ($rowNumber === '' || ($description === '' && $amount === null && $notes === '')) {
                    continue;
                }

                $suspectedAmountNote = null;
                $noteAmount = $this->parseInt($notes);
                if ($noteAmount !== null && ($amount === null || $amount === 0)) {
                    $suspectedAmountNote = $noteAmount;
                    $this->warn($file, $sheet, $row, 'یک مبلغ احتمالی در ستون توضیحات پیدا شد.', [
                        'description' => $description,
                        'amount_column' => $amount,
                        'notes_column' => $notes,
                    ]);
                }

                $this->insert('transactions', [
                    'batch_id' => $batchId,
                    'row_number' => $this->parseInt($rowNumber),
                    'invoice_count' => $this->clean($reader->value($sheet, "B{$row}")),
                    'tx_date' => $this->composeJalaliDate(
                        $reader->value($sheet, "C{$row}"),
                        $reader->value($sheet, "D{$row}"),
                        $reader->value($sheet, "E{$row}")
                    ),
                    'description' => $description,
                    'amount' => $amount,
                    'notes' => $notes,
                    'category' => $this->categorizeTransaction($description, $amount),
                    'suspected_amount_note' => $suspectedAmountNote,
                ]);
            }
        }
    }

    private function insert(string $table, array $data): void
    {
        $columns = array_keys($data);
        $placeholders = array_map(static fn (string $column): string => ':' . $column, $columns);
        $sql = sprintf(
            'INSERT INTO %s (%s) VALUES (%s)',
            $table,
            implode(', ', $columns),
            implode(', ', $placeholders)
        );
        $statement = $this->pdo->prepare($sql);
        foreach ($data as $column => $value) {
            $statement->bindValue(':' . $column, $value);
        }
        $statement->execute();
    }

    private function warn(string $file, string $sheet, int $row, string $message, array $payload = []): void
    {
        $this->insert('import_warnings', [
            'file_name' => $file,
            'sheet_name' => $sheet,
            'row_number' => $row,
            'message' => $message,
            'payload' => json_encode($payload, JSON_UNESCAPED_UNICODE),
        ]);
    }

    private function count(string $table): int
    {
        return (int) $this->pdo->query("SELECT COUNT(*) FROM {$table}")->fetchColumn();
    }

    private function clean(mixed $value): string
    {
        if ($value === null) {
            return '';
        }
        $text = trim((string) $value);
        return strtr($text, [
            '۰' => '0', '۱' => '1', '۲' => '2', '۳' => '3', '۴' => '4',
            '۵' => '5', '۶' => '6', '۷' => '7', '۸' => '8', '۹' => '9',
            '٠' => '0', '١' => '1', '٢' => '2', '٣' => '3', '٤' => '4',
            '٥' => '5', '٦' => '6', '٧' => '7', '٨' => '8', '٩' => '9',
        ]);
    }

    private function parseInt(mixed $value): ?int
    {
        $text = str_replace([',', '٬'], '', $this->clean($value));
        if ($text === '' || $text === '-') {
            return null;
        }
        if (!preg_match('/^-?\d+(\.0+)?$/', $text)) {
            return null;
        }

        return (int) (float) $text;
    }

    private function parseJalaliDate(mixed $value): string
    {
        $text = $this->clean($value);
        if ($text === '' || $text === '-') {
            return '';
        }
        if (!preg_match('/^(\d{2,4})\/(\d{1,2})\/(\d{1,2})$/', $text, $matches)) {
            return $text;
        }

        return sprintf('%04d/%02d/%02d', (int) $matches[1], (int) $matches[2], (int) $matches[3]);
    }

    private function composeJalaliDate(mixed $day, mixed $month, mixed $year): string
    {
        $d = $this->parseInt($day);
        $m = $this->parseInt($month);
        $y = $this->parseInt($year);
        if ($d === null || $m === null || $y === null) {
            return '';
        }

        return sprintf('%04d/%02d/%02d', $y, $m, $d);
    }

    private function isValidJalaliDate(string $date): bool
    {
        if (!preg_match('/^(\d{4})\/(\d{2})\/(\d{2})$/', $date, $matches)) {
            return true;
        }
        $month = (int) $matches[2];
        $day = (int) $matches[3];
        if ($month < 1 || $month > 12) {
            return false;
        }
        $maxDay = $month <= 6 ? 31 : 30;
        if ($month === 12) {
            $maxDay = 29;
        }

        return $day >= 1 && $day <= $maxDay;
    }

    private function categorizeTransaction(string $description, ?int $amount): string
    {
        if ($amount === null) {
            return 'نامشخص';
        }
        if ($amount < 0) {
            return 'هزینه';
        }
        foreach (['آبونمان', 'شارژ', 'سود', 'سهم مرکز', 'واریز'] as $keyword) {
            if (str_contains($description, $keyword)) {
                return 'درآمد';
            }
        }

        return 'دریافت';
    }
}

File: app/importer.py
from __future__ import annotations

import json
import re
import sqlite3
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from openpyxl import load_workbook
from openpyxl.worksheet.worksheet import Worksheet

from .database import init_db, reset_imported_data


MONTHS = [
    ("E", 1, "فروردین"),
    ("F", 2, "اردیبهشت"),
    ("G", 3, "خرداد"),
    ("H", 4, "تیر"),
    ("I", 5, "مرداد"),
    ("J", 6, "شهریور"),
    ("K", 7, "مهر"),
    ("L", 8, "آبان"),
    ("M", 9, "آذر"),
    ("N", 10, "دی"),
    ("O", 11, "بهمن"),
    ("P", 12, "اسفند"),
]

PERSIAN_DIGITS = str.maketrans("۰۱۲۳۴۵۶۷۸۹٠١٢٣٤٥٦٧٨٩", "01234567890123456789")


@dataclass(frozen=True)
class WorkbookSource:
    path: Path
    name: str


def clean(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, float) and value.is_integer():
        value = int(value)
    return str(value).translate(PERSIAN_DIGITS).strip()


def parse_int(value: Any) -> int | None:
    text = clean(value)
    if not text or text == "-":
        return None
    text = text.replace(",", "").replace("٬", "")
    if re.fullmatch(r"-?\d+(\.0+)?", text):
        return int(float(text))
    return None


def parse_amount(value: Any) -> int | None:
    return parse_int(value)


def parse_jalali_date(value: Any) -> str:
    text = clean(value)
    if not text or text == "-":
        return ""
    match = re.fullmatch(r"(\d{2,4})/(\d{1,2})/(\d{1,2})", text)
    if not match:
        return text
    year, month, day = (int(part) for part in match.groups())
    return f"{year:04d}/{month:02d}/{day:02d}"


def compose_jalali_date(day: Any, month: Any, year: Any) -> str:
    d = parse_int(day)
    m = parse_int(month)
    y = parse_int(year)
    if d is None or m is None or y is None:
        return ""
    return f"{y:04d}/{m:02d}/{d:02d}"


def is_valid_jalali_date(date_text: str) -> bool:
    match = re.fullmatch(r"(\d{4})/(\d{2})/(\d{2})", date_text)
    if not match:
        return True
    _, month, day = (int(part) for part in match.groups())
    if month < 1 or month > 12:
        return False
    max_day = 31 if month <= 6 else 30
    if month == 12:
        max_day = 29
    return 1 <= day <= max_day


def cell_reader(sheet: Worksheet):
    merged_lookup: dict[str, str] = {}
    for merged_range in sheet.merged_cells.ranges:
        anchor = merged_range.start_cell.coordinate
        for row in sheet[merged_range.coord]:
            for cell in row:
                merged_lookup[cell.coordinate] = anchor

    def value(coordinate: str) -> Any:
        anchor = merged_lookup.get(coordinate, coordinate)
        return sheet[anchor].value

    return value


def warn(
    conn: sqlite3.Connection,
    file_name: str,
    sheet_name: str,
    row_number: int | None,
    message: str,
    payload: dict[str, Any] | None = None,
) -> None:
    conn.execute(
        """
        INSERT INTO import_warnings (file_name, sheet_name, row_number, message, payload)
        VALUES (?, ?, ?, ?, ?)
        """,
        (
            file_name,
            sheet_name,
            row_number,
            message,
            json.dumps(payload or {}, ensure_ascii=False),
        ),
    )


def import_innovation_center(conn: sqlite3.Connection, source: WorkbookSource) -> None:
    workbook = load_workbook(source.path, data_only=True)

    members_sheet = workbook["Members"]
    get = cell_reader(members_sheet)
    for row in range(4, members_sheet.max_row + 1):
        full_name = clean(get(f"C{row}"))
        if not full_name:
            continue
        conn.execute(
            """
            INSERT INTO members
                (row_number, code, full_name, team_name, desks, lockers, power_strips,
                 phone, national_id, notes, source_file, source_sheet)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                row,
                clean(get(f"B{row}")),
                full_name,
                clean(get(f"D{row}")),
                clean(get(f"E{row}")),
                clean(get(f"F{row}")),
                clean(get(f"G{row}")),
                clean(get(f"H{row}")),
                clean(get(f"I{row}")),
                clean(get(f"J{row}")),
                source.name,
                members_sheet.title,
            ),
        )

    teams_sheet = workbook["Teams"]
    get = cell_reader(teams_sheet)
    for row in range(6, 20):
        name = clean(get(f"B{row}"))
        leader = clean(get(f"C{row}"))
        if not name and not leader:
            continue
        conn.execute(
            """
            INSERT INTO teams
                (row_number, name, leader, phone, desk_count, lockers, power_strips,
                 joined_at, warning, notes, source_file, source_sheet)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                row,
                name,
                leader,
                clean(get(f"D{row}")),
                parse_int(get(f"E{row}")),
                clean(get(f"F{row}")),
                clean(get(f"G{row}")),
                parse_jalali_date(get(f"H{row}")),
                clean(get(f"I{row}")),
                clean(get(f"J{row}")),
                source.name,
                teams_sheet.title,
            ),
        )

    lockers_sheet = workbook["lockers"]
    get = cell_reader(lockers_sheet)
    for row in range(6, lockers_sheet.max_row + 1):
        locker_number = parse_int(get(f"A{row}"))
        if locker_number is None:
            continue
        delivered_at = parse_jalali_date(get(f"D{row}"))
        if delivered_at and not is_valid_jalali_date(delivered_at):
            warn(
                conn,
                source.name,
                lockers_sheet.title,
                row,
                "تاریخ تحویل کمد معتبر به نظر نمی‌رسد.",
                {"delivered_at": delivered_at, "locker_number": locker_number},
            )
        conn.execute(
            """
            INSERT INTO lockers
                (locker_number, status, assigned_to, delivered_at, key_number,
                 spare_key, notes, source_file, source_sheet)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                locker_number,
                clean(get(f"B{row}")),
                clean(get(f"C{row}")),
                delivered_at,
                clean(get(f"E{row}")),
                clean(get(f"F{row}")),
                "",
                source.name,
                lockers_sheet.title,
            ),
        )

    plans_sheet = workbook["plans"]
    get = cell_reader(plans_sheet)
    for row in range(6, plans_sheet.max_row + 1):
        plan_number = parse_int(get(f"A{row}"))
        title = clean(get(f"C{row}"))
        if plan_number is None and not title:
            continue
        conn.execute(
            """
            INSERT INTO plans
                (plan_number, status, title, proposed_budget, cost_type, schedule,
                 notes, source_file, source_sheet)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                plan_number,
                clean(get(f"B{row}")),
                title,
                parse_amount(get(f"D{row}")),
                clean(get(f"E{row}")),
                clean(get(f"F{row}")),
                "",
                source.name,
                plans_sheet.title,
            ),
        )


def import_charges(conn: sqlite3.Connection, source: WorkbookSource) -> None:
    workbook = load_workbook(source.path, data_only=True)
    for sheet in workbook.worksheets:
        get = cell_reader(sheet)
        charge_rate = parse_amount(get("T1"))
        rent_rate = parse_amount(get("T3"))
        for row in range(6, sheet.max_row + 1):
            team_name = clean(get(f"B{row}"))
            leader = clean(get(f"C{row}"))
            if not team_name and not leader:
                continue
            if clean(get(f"A{row}")) and not parse_int(get(f"A{row}")):
                continue
            note = clean(get(f"Q{row}"))
            for column, month_index, month_name in MONTHS:
                raw_amount = get(f"{column}{row}")
                amount = parse_amount(raw_amount)
                if amount is None:
                    continue
                conn.execute(
                    """
                    INSERT INTO charges
                        (fiscal_year, team_name, leader, desk_count, month_index,
                         month_name, amount, note, charge_rate, rent_rate,
                         source_file, source_sheet)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        sheet.title,
                        team_name,
                        leader,
                        parse_int(get(f"D{row}")),
                        month_index,
                        month_name,
                        amount,
                        note,
                        charge_rate,
                        rent_rate,
                        source.name,
                        sheet.title,
                    ),
                )


def categorize_transaction(description: str, amount: int | None) -> str:
    if amount is None:
        return "نامشخص"
    if amount < 0:
        return "هزینه"
    keywords = ("آبونمان", "شارژ", "سود", "سهم مرکز", "واریز")
    if any(keyword in description for keyword in keywords):
        return "درآمد"
    return "دریافت"


def import_finance(conn: sqlite3.Connection, source: WorkbookSource) -> None:
    workbook = load_workbook(source.path, data_only=True)
    for sheet in workbook.worksheets:
        get = cell_reader(sheet)
        conn.execute(
            """
            INSERT INTO financial_batches
                (sheet_name, petty_cash_holder, petty_cash_number, previous_balance,
                 new_deposit, total_balance, received_at, from_date, to_date, source_file)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                sheet.title,
                clean(get("I1")),
                clean(get("K1")),
                parse_amount(get("I2")),
                parse_amount(get("I3")),
                parse_amount(get("I4")),
                parse_jalali_date(get("K2")),
                parse_jalali_date(get("K3")),
                parse_jalali_date(get("K4")),
                source.name,
            ),
        )
        batch_id = conn.execute("SELECT last_insert_rowid()").fetchone()[0]

        for row in range(6, 28):
            row_number = clean(get(f"A{row}"))
            description = clean(get(f"F{row}"))
            amount = parse_amount(get(f"I{row}"))
            notes = clean(get(f"J{row}"))
            suspected_amount_note = None
            if row_number and (description or amount is not None or notes):
                note_amount = parse_amount(notes)
                if note_amount is not None and (amount is None or amount == 0):
                    suspected_amount_note = note_amount
                    warn(
                        conn,
                        source.name,
                        sheet.title,
                        row,
                        "یک مبلغ احتمالی در ستون توضیحات پیدا شد.",
                        {
                            "description": description,
                            "amount_column": amount,
                            "notes_column": notes,
                        },
                    )
                conn.execute(
                    """
                    INSERT INTO transactions
                        (batch_id, row_number, invoice_count, tx_date, description,
                         amount, notes, category, suspected_amount_note)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        batch_id,
                        parse_int(row_number),
                        clean(get(f"B{row}")),
                        compose_jalali_date(get(f"C{row}"), get(f"D{row}"), get(f"E{row}")),
                        description,
                        amount,
                        notes,
                        categorize_transaction(description, amount),
                        suspected_amount_note,
                    ),
                )


def import_all(conn: sqlite3.Connection, base_dir: str | Path) -> None:
    base = Path(base_dir)
    init_db(conn)
    reset_imported_data(conn)
    sources = {
        "innovation": WorkbookSource(base / "Innovation Center.xlsx", "Innovation Center.xlsx"),
        "charges": WorkbookSource(base / "CHARGE.xlsx", "CHARGE.xlsx"),
        "finance": WorkbookSource(base / "finance.xlsx", "finance.xlsx"),
    }
    for source in sources.values():
        if not source.path.exists():
            raise FileNotFoundError(f"Required workbook not found: {source.path}")

    import_innovation_center(conn, sources["innovation"])
    import_charges(conn, sources["charges"])
    import_finance(conn, sources["finance"])
    conn.execute(
        "INSERT INTO import_runs (source_files) VALUES (?)",
        (json.dumps([source.name for source in sources.values()], ensure_ascii=False),),
    )
    conn.commit()
